crates/analyzer/src/traversal/expressions.rs

```diff
@@ -873,15 +880,73 @@ fn check_for_call_to_init_fn(
     }
 }
 
+fn check_for_self_param(scope: &mut BlockScope, use_span: Span) -> Result<(), FatalError> {
+    if scope.root.function.signature(scope.db()).self_decl == SelfDecl::None {
+        scope.fancy_error(
+            "`self` is not defined",
+            vec![Label::primary(use_span, "undefined value")],
+            vec![
+                "add `self` to the scope by including it in the function signature".to_string(),
+                "Example: `fn foo(self, bar: bool)`".to_string(),
+            ],
+        );
+        Err(FatalError::new())
```

Comment:
I think just adding the error should be enough and the function doesn't need to return anything. That way we could get rid of the `FatalError` which is preventing us from collecting more errors.